// lobby/lib/appBase2.js
'use strict'

var events = require('events');
var pather = require("path");
var util = require('util');
var log4js = require('log4js');
var isobject = require('isobject');

var AppBase = function(name){

    this._appName = name;
    this._handlers = {};

    this.addArgs = (handlerArgs, args) => {
        this.log("adding args" + args);
        for(var n in args){
            if(!isobject(handlerArgs[n])){
                handlerArgs[n] = args[n];
                this.log(`adding arg[${args[n]}]`);
            }
        }
        this.log(`handleArgs[${handlerArgs}]`);
    };
}
util.inherits(AppBase, events.EventEmitter);

// AppBase.prototype._handlers = {};

AppBase.prototype.log = function(str){
    console.log("[" + this._appName + "]" + str);
};

AppBase.prototype.setStatus = function(status){

    this._status = status;

    this.log("Status Change:" + this._status);

    var args = this._handlers[status];

    this.emit(status, args);

    // this.doEvent(AppBase.Handles.onStatusChanging);
};

AppBase.prototype.registerEvent = function(status, handler, args){
    var handleArgs = this._handlers[status];

    this.log(handler);
    this.log(args);

    if(handleArgs && handleArgs instanceof Object){

        // this.addArgs(handleArgs, args);
    }else{
        handleArgs = {};
        this._handlers[status] = handleArgs;
    }

    if(args){
        this.addArgs(handleArgs, args);
    }

    this.on(status, handler);
};

module.exports = AppBase;

// var a = new AppBase("testerA");
// var h = function(args){
//     console.log(args);
// }
// var as = {name:"neo"};
// a.registerEvent("1", h, as);
// a.setStatus("1");

// var WebApp = function(name){
//     AppBase.call(this,name);
// }
// util.inherits(WebApp, AppBase);

// var w = new WebApp("webapp");
// var hw = function(args){
//     console.log(args);
// }
// var aw = {name:"neo"};
// w.registerEvent("1", hw, aw);
// w.setStatus("1");

// var WebApp2 = function(name, name1){
//     AppBase.call(this,name1);
// }
// util.inherits(WebApp2, AppBase);

// var w2 = new WebApp2("webapp2","webapp3");
// // var hw = function(args){
// //     console.log(args);
// // }
// // var aw = {name:"neo"};
// w2.registerEvent("1", hw, aw);
// w2.setStatus("1");

// console.log(a.test1);
// a.test1();

// var WebApp = function(app, path, name){
//     this.app = app;
//     this.path = path;
//     this.name = name;
// }
// util.inherits(WebApp, AppBase);

// var obj = new WebApp("app", "path", "name");
// console.log(obj);
